### PR TITLE
Simplify the synth machine, added quickstart nb + mock executor

### PR DESCRIPTION
**Pull request by simonbutt**

### CODE REVIEW

**Comment by simonbutt** [comment]
> No notebooks. Write docs.

We need both. For python developers, notebooks are an easy way to  get up and running with a new framework.

I'll remove from this change and convert to colab, so that we don't need to add jupyter to dev environments.

File: synth_machine/machine.py [hunk 1]
```diff
@@ -590,3 +597,38 @@ async def streaming_trigger(self, trigger: str, params: Optional[dict] = None):
...
+        else:
+            raise Exception(
+                f"No transition: {trigger} exists at state: {self.current_state()}"
+            )
+        missing_inputs = []
+        for input in filtered_transition[0].get("inputs", []):
+            if input["key"] not in (self.memory | inputs).keys():
+                missing_inputs.append(input["key"])
+        if len(missing_inputs) > 0:
+            raise Exception(
```
Comment (dark0dave) [inline synth_machine/machine.py:625]: give a better exception not base exceptio

**Comment by simonbutt** [comment]
- synth validate as pydantic
- Tool into init.py
- fake executor -> mock
- check if key exists, if exists, then import

**Comment by simonbutt** [comment]
> * synth validate as pydantic
> * Tool into init.py
> * fake executor -> mock
> * check if key exists, if exists, then import

Remove notebook for this PR

File: synth_machine/runners.py [hunk 2]
```diff
@@ -18,9 +18,15 @@ async def tool_runner(
             config["tool_path"],
             json=config["payload"],
         )
+        response_meta = {
```
Comment (simonbutt) [inline synth_machine/runners.py:21]: response_headers

File: synth_machine/runners.py [hunk 3]
```diff
@@ -18,9 +18,15 @@ async def tool_runner(
...
+            "tool_meta": {
+                "status": response.status_code,
+                "success": str(response.status_code).startswith("2"),
```
Comment (simonbutt) [inline synth_machine/runners.py:24]: Less hacky way to approach

File: synth_machine/executors/__init__.py [hunk 4]
```diff
@@ -1,7 +1,6 @@
 import os
 
 DEBUG = os.environ.get("DEBUG")
-ANTHROPIC_API_KEY = os.environ.get("ANTHROPIC_API_KEY")
-OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")
-TOGETHER_API_KEY = os.environ.get("TOGETHER_API_KEY")
-VERTEX_REGION = os.environ.get("VERTEX_REGION", "us-central1")
+ANTHROPIC_API_KEY = os.environ.get("ANTHROPIC_API_KEY", "MOCK_KEY")
```
Comment (dark0dave) [inline synth_machine/executors/__init__.py:4]: Remove Mock Key

File: synth_machine/config.py [hunk 5]
```diff
@@ -23,16 +22,19 @@ class SynthConfig:
 
 
 async def tool_setup(
-    output_definition: dict, inputs: dict, id: str, user: str, tools: list
+    output_definition: Output, inputs: dict, id: str, user: str, tools: list
```
Comment (dark0dave) [inline synth_machine/config.py:25]:
```suggestion
    output_definition: Output, inputs: Input, id: str, user: str, tools: list
```

File: synth_machine/config.py [hunk 6]
```diff
@@ -23,16 +22,19 @@ class SynthConfig:
...
 ) -> dict:
-    tool_search = [tool for tool in tools if tool.name == output_definition.get("tool")]
+    tool_search = [tool for tool in tools if tool.name == output_definition.tool]
     if not tool_search:
+        logging.warning(
```
Comment (dark0dave) [inline synth_machine/config.py:29]:
```suggestion
        logging.error(
```

File: synth_machine/config.py [hunk 7]
```diff
@@ -101,17 +102,24 @@ async def prompt_setup(
...
-        return (None, system_err)
+    system_prompt_template = output_definition.system_prompt
+    if system_prompt_template:
+        system_prompt, system_err = prompt_for_transition(
+            inputs=inputs,
+            prompt_template=system_prompt_template,
+        )
+        if system_err:
+            return (None, system_err)
+    else:
```
Comment (dark0dave) [inline synth_machine/config.py:113]: this else is not needed

File: synth_machine/executor_factory.py [hunk 8]
```diff
@@ -2,11 +2,17 @@
...
+import os
 
 
 def get_executor(name: str) -> BaseExecutor:
-    return {
-        "openai": OpenAIExecutor(),
-        "togetherai": TogetherAIExecutor(),
-        "anthropic": AnthropicExecutor(),
-    }[name]
+    base_executor_runners: Dict[str, BaseExecutor] = {"fake": FakeExecutor()}
```
Comment (dark0dave) [inline synth_machine/executor_factory.py]: maybe define this outside of the function, so its not recreated each time?

File: synth_machine/executors/fake.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    calculate_input_tokens,
+)
+import random
+import asyncio
+
+
+@singleton
+class FakeExecutor(BaseExecutor):
+    def __init__(self) -> None:
+        self.word_catalog = [
```
Comment (dark0dave) [inline synth_machine/executors/fake.py]: is this really needed? create a fixture if you really want this. This should be basically empty.

File: synth_machine/executor_factory.py [hunk 10]
```diff
@@ -2,11 +2,17 @@
...
+import os
 
 
 def get_executor(name: str) -> BaseExecutor:
-    return {
-        "openai": OpenAIExecutor(),
-        "togetherai": TogetherAIExecutor(),
-        "anthropic": AnthropicExecutor(),
-    }[name]
+    base_executor_runners: Dict[str, BaseExecutor] = {"lorem": LoremExecutor()}
```
Comment (dark0dave) [inline synth_machine/executor_factory.py:11]: Do you want to move this outside of here?

**Review by dark0dave — APPROVED** [review]

**Pull request merged**